packages/backend/src/open_payments/invoice/service.ts

```diff
@@ -110,62 +113,146 @@ async function handleInvoicePayment(
   if (!amountReceived) {
     return
   }
-  const invoice = await Invoice.query(deps.knex)
+  await Invoice.query(deps.knex)
     .patch({
-      active: false
+      active: false,
+      webhookId: uuid()
     })
     .where('id', invoiceId)
     .andWhere('amount', '<=', amountReceived.toString())
-    .returning('*')
-    .first()
-  if (invoice) {
-    await deps.webhookService.send({
-      type: EventType.InvoicePaid,
-      invoice,
-      amountReceived
-    })
-  }
 }
 
-// Deactivate expired invoices that have some money.
-// Delete expired invoices that have never received money.
+// Fetch (and lock) an invoice for work.
 // Returns the id of the processed invoice (if any).
-async function deactivateNextInvoice(
-  deps: ServiceDependencies
+async function processNextInvoice(
+  deps_: ServiceDependencies
 ): Promise<string | undefined> {
-  return deps.knex.transaction(async (trx) => {
+  return deps_.knex.transaction(async (trx) => {
+    const now = new Date(Date.now()).toISOString()
     // 30 seconds backwards to allow a prepared (but not yet fulfilled/rejected) packet to finish before being deactivated.
-    const now = new Date(Date.now() - 30_000).toISOString()
+    const delayedNow = new Date(Date.now() - 30_000).toISOString()
     const invoices = await Invoice.query(trx)
       .limit(1)
       // Ensure the invoices cannot be processed concurrently by multiple workers.
       .forUpdate()
       // If an invoice is locked, don't wait — just come back for it later.
       .skipLocked()
-      .where('active', true)
-      .andWhere('expiresAt', '<', now)
+      .where((builder: QueryBuilder) => {
+        builder.where('active', true).andWhere('expiresAt', '<', delayedNow)
+      })
```

Comment:
The index on `expiresAt` should be updated to include `active`. And include a `TODO` that once knex/objection support partial indexes we should switch to indexing it on just `active:true`.